std/src/main/scala/algebra/std/map.scala

```diff
@@ -1,58 +1,132 @@
 package algebra
 package std
 
-import algebra.ring.{AdditiveMonoid, Rng}
+import algebra.ring._
+import algebra.std.util.StaticMethods.{addMap, initMutableMap, subtractMap, wrapMutableMap}
+import scala.collection.mutable
 
 package object map extends MapInstances
 
-trait MapInstances {
-  implicit def mapEq[K, V: Eq: AdditiveMonoid]: Eq[Map[K, V]] =
-    new MapEq[K, V]
-  
+trait MapInstances extends MapInstances3
+
+trait MapInstances3 extends MapInstances2 {
   implicit def mapRng[K, V: Rng] =
     new MapRng[K, V]
 }
 
-class MapEq[K, V](implicit ev: Eq[V], a: AdditiveMonoid[V]) extends Eq[Map[K, V]] {
+trait MapInstances2 extends MapInstances1 {
+  implicit def mapSemiring[K, V: Semiring]: MapSemiring[K, V] =
+    new MapSemiring[K, V]
+}
+
+trait MapInstances1 extends MapInstances0 {
+  implicit def mapGroup[K, V: Group]: MapGroup[K, V] =
+    new MapGroup[K, V]
+  implicit def mapAdditiveGroup[K, V: AdditiveGroup]: MapAdditiveGroup[K, V] =
+    new MapAdditiveGroup[K, V]
+}
+
+trait MapInstances0 {
+  implicit def mapEq[K, V: Eq: AdditiveMonoid]: Eq[Map[K, V]] =
```

Comment:
what about cases where we have an AdditiveMonoid, e.g. Int, but want an exact Eq (given by MapEq). I guess, in that case, you can `import std.map._` but then `implicit def mapEq[K, V: Eq]: Eq[Map[K, V]] = new MapEq`, and that will work? (does the local implicit in this case beat the imported implicit?)